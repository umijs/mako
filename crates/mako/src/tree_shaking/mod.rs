pub mod analyze_statement;
pub mod defined_ident_collector;
pub mod module_side_effects_flag;
pub mod statement;
pub mod statement_graph;
#[allow(clippy::module_inception)]
pub mod tree_shaking;
pub mod tree_shaking_analyze;
pub mod tree_shaking_module;
pub mod unused_statement_cleanup;
pub mod unused_statement_marker;
pub mod unused_statement_sweep;
pub mod used_ident_collector;
