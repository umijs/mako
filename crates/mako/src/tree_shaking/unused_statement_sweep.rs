use mako_core::swc_common::util::take::Take;
use mako_core::swc_ecma_ast;
use mako_core::swc_ecma_ast::{Decl, ExportDecl, ModuleItem};
use mako_core::swc_ecma_visit::{VisitMut, VisitMutWith};
use mako_core::tracing::debug;

use crate::comments::Comments;
use crate::module::ModuleId;

pub struct UnusedStatementSweep<'a> {
    module_id: ModuleId,
    pub comments: &'a Comments,
    need_removed_module_item: bool,
    pub removed_item_count: usize,
}

impl<'a> UnusedStatementSweep<'a> {
    pub fn new(module_id: &ModuleId, comments: &'a Comments) -> Self {
        Self {
            module_id: module_id.clone(),
            comments,
            need_removed_module_item: false,
            removed_item_count: 0,
        }
    }
}

impl VisitMut for UnusedStatementSweep<'_> {
    fn visit_mut_module(&mut self, n: &mut swc_ecma_ast::Module) {
        if self.comments.has_unused_module(n.span) {
            debug!("remove module {}", self.module_id.id);
            n.take();
        } else {
            n.visit_mut_children_with(self);
        }
    }

    fn visit_mut_module_item(&mut self, decl: &mut ModuleItem) {
        let old_removed_module_item = self.need_removed_module_item;
        self.need_removed_module_item = false;

        decl.visit_mut_children_with(self);

        if self.need_removed_module_item {
            debug!("remove module item: {:?}", decl);
            decl.take();
        }
        self.need_removed_module_item = old_removed_module_item;
    }

    // TODO: 目前 jsx 会被删除掉，等待解决
    fn visit_mut_import_decl(&mut self, import_decl: &mut swc_ecma_ast::ImportDecl) {
        let mut removed = vec![];
        for (index, specifier) in import_decl.specifiers.iter().enumerate() {
            if let swc_ecma_ast::ImportSpecifier::Named(named_specifier) = specifier {
                if self.comments.has_unused(named_specifier.span) {
                    removed.push(index);
                }
            }
        }
        removed.reverse();
        for index in removed {
            import_decl.specifiers.remove(index);
            self.removed_item_count += 1;
        }
        if import_decl.specifiers.is_empty() {
            self.need_removed_module_item = true;
        }
    }

    fn visit_mut_export_specifiers(&mut self, specifiers: &mut Vec<swc_ecma_ast::ExportSpecifier>) {
        let mut removed = vec![];
        for (index, specifier) in specifiers.iter().enumerate() {
            if let swc_ecma_ast::ExportSpecifier::Named(named_specifier) = specifier {
                if self.comments.has_unused(named_specifier.span) {
                    removed.push(index);
                }
            }
        }
        removed.reverse();
        for index in removed {
            specifiers.remove(index);
            self.removed_item_count += 1;
        }
        if specifiers.is_empty() {
            self.need_removed_module_item = true;
        }
    }

    fn visit_mut_export_default_expr(
        &mut self,
        default_expr: &mut swc_ecma_ast::ExportDefaultExpr,
    ) {
        if self.comments.has_unused(default_expr.span) {
            self.need_removed_module_item = true;
            self.removed_item_count += 1;
        }
    }

    fn visit_mut_export_default_decl(
        &mut self,
        default_decl: &mut swc_ecma_ast::ExportDefaultDecl,
    ) {
        if self.comments.has_unused(default_decl.span) {
            self.need_removed_module_item = true;
            self.removed_item_count += 1;
        }
    }

    fn visit_mut_export_decl(&mut self, export_decl: &mut ExportDecl) {
        match &mut export_decl.decl {
            Decl::Var(var_decl) => {
                self.remove_unused_decls(var_decl);
                if var_decl.decls.is_empty() {
                    self.need_removed_module_item = true;
                }
            }
            Decl::Class(class_decl) => {
                if self.comments.has_unused(class_decl.ident.span) {
                    self.need_removed_module_item = true;
                    self.removed_item_count += 1;
                }
            }
            Decl::Fn(fn_decl) => {
                if self.comments.has_unused(fn_decl.ident.span) {
                    self.need_removed_module_item = true;
                    self.removed_item_count += 1;
                }
            }
            _ => (),
        }
    }
}

impl UnusedStatementSweep<'_> {
    fn remove_unused_decls(&mut self, var_decl: &mut Box<swc_ecma_ast::VarDecl>) {
        let mut removed = vec![];
        for (index, decl) in var_decl.decls.iter().enumerate() {
            if self.comments.has_unused(decl.span) {
                removed.push(index);
            }
        }
        removed.reverse();
        for index in removed {
            var_decl.decls.remove(index);
            self.removed_item_count += 1;
        }
    }
}
