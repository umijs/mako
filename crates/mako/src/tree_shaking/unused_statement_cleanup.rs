use swc_ecma_ast::{ModuleDecl, ModuleItem};
use swc_ecma_visit::VisitMut;
use tracing::debug;

use crate::module::ModuleId;
use crate::tree_shaking::tree_shaking_module::{should_skip, UsedIdentHashMap};

pub struct UnusedStatementCleanup<'a> {
    module_id: ModuleId,
    used_export_statement: &'a UsedIdentHashMap,
}

impl<'a> UnusedStatementCleanup<'a> {
    pub fn new(module_id: &ModuleId, used_export_statement: &'a UsedIdentHashMap) -> Self {
        Self {
            module_id: module_id.clone(),
            used_export_statement,
        }
    }
}

impl VisitMut for UnusedStatementCleanup<'_> {
    fn visit_mut_module(&mut self, module: &mut swc_ecma_ast::Module) {
        if should_skip(&self.module_id.id) {
            return;
        }
        let mut removed_ids = vec![];
        for (id, statement) in module.body.iter().enumerate() {
            if let ModuleItem::ModuleDecl(ModuleDecl::Import(import_decl)) = statement {
                if should_skip(&import_decl.src.value) {
                    continue;
                }
            }
            if !self.used_export_statement.has_stmt_by_id(&id) {
                removed_ids.push(id);
            }
        }

        removed_ids.sort();
        removed_ids.reverse();
        for id in &removed_ids {
            module.body.remove(*id);
        }

        debug!(
            "reexport statement cleanup {:?} {:?} {:?}",
            &self.module_id, &removed_ids, &self.used_export_statement
        );
    }
}
