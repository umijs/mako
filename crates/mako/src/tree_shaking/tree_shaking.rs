use std::collections::HashMap;

use mako_core::swc_ecma_visit::VisitMutWith;
use mako_core::tracing::debug;

use crate::compiler::Compiler;
use crate::module::ModuleId;
use crate::tree_shaking::tree_shaking_module::{ModuleSystem, TreeShakingModule, UsedExports};
use crate::tree_shaking::unused_statement_cleanup::UnusedStatementCleanup;
use crate::tree_shaking::unused_statement_marker::UnusedStatementMarker;
use crate::tree_shaking::unused_statement_sweep::UnusedStatementSweep;

impl Compiler {
    pub fn tree_shaking(&self) -> Vec<ModuleId> {
        // 拓扑排序
        let (entry_modules, sorted_modules, cycle_modules) = self.make_toposort();
        debug!("entry_modules: {:?}", &entry_modules);
        debug!("cycle_modules: {:?}", &cycle_modules);

        // 入口模块设置为副作用模块
        self.markup_entry_modules_as_side_effects(entry_modules);

        // 循环依赖模块设置为副作用
        // 场景比如：在 app.ts 里使用 umi 的 createGlobalStyle 时
        self.markup_cycle_modules_as_side_effects(cycle_modules);

        let (tree_shaking_module_ids, mut tree_shaking_module_map) =
            self.create_tree_shaking_module_map(sorted_modules);

        let tree_shaking_module_map = &mut tree_shaking_module_map;

        let mut modules_to_remove = vec![];
        for tree_shaking_module_id in &tree_shaking_module_ids {
            let tree_shaking_module = tree_shaking_module_map
                .get_mut(tree_shaking_module_id)
                .unwrap();
            let module_graph = self.context.module_graph.read().unwrap();
            let module = module_graph.get_module(&tree_shaking_module.id).unwrap();
            let module_is_side_effects = module.info.as_ref().unwrap().get_side_effects_flag();
            drop(module_graph);
            debug!(
                "tree_shaking module: [side_effects:{}/{}] {}",
                module_is_side_effects,
                tree_shaking_module.side_effects,
                &tree_shaking_module_id.id,
            );
            debug!(
                "    - used_exports: {:?}",
                &tree_shaking_module.used_exports
            );

            if matches!(tree_shaking_module.module_system, ModuleSystem::ESModule) {
                // 包含副作用的模块
                if tree_shaking_module.side_effects {
                    {
                        {
                            let imports = tree_shaking_module.imports();
                            let exports = tree_shaking_module.exports();

                            // 分析使用情况
                            for import in imports {
                                debug!("    - import: {:?}", &import);
                                self.analyze_import_statement(
                                    tree_shaking_module_map,
                                    tree_shaking_module_id,
                                    import,
                                );
                            }

                            for export in exports {
                                debug!("    - export: {:?}", &export);
                                self.analyze_export_statement(
                                    tree_shaking_module_map,
                                    tree_shaking_module_id,
                                    export,
                                );
                            }
                        };
                    }
                } else {
                    // 不包含副作用的模块，执行 tree shaking
                    let tree_shaking_module = tree_shaking_module_map
                        .get_mut(tree_shaking_module_id)
                        .unwrap();

                    let mut module_graph = self.context.module_graph.write().unwrap();
                    let module = module_graph
                        .get_module_mut(&tree_shaking_module.id)
                        .unwrap();
                    let ast = module.info.as_mut().unwrap().ast.as_script_mut();

                    let unused = tree_shaking_module.get_used_export_statement();
                    // 把没有用到的 stmt 删除
                    let mut reexport_cleanup =
                        UnusedStatementCleanup::new(&tree_shaking_module.id, &unused);
                    ast.visit_mut_with(&mut reexport_cleanup);

                    // 通过 tree_shaking_module 进行无用的标记
                    let mut comments = self.context.meta.script.output_comments.write().unwrap();
                    let mut marker =
                        UnusedStatementMarker::new(tree_shaking_module, unused, &mut comments);
                    ast.visit_mut_with(&mut marker);

                    let mut sweep = UnusedStatementSweep::new(&tree_shaking_module.id, &comments);
                    ast.visit_mut_with(&mut sweep);
                    tree_shaking_module.update_statement(module);

                    drop(module_graph);
                    drop(comments);

                    // 当前模块没有被使用到的导出，删除当前模块
                    if tree_shaking_module.used_exports.is_empty() {
                        modules_to_remove.push(tree_shaking_module_id.clone());
                        continue;
                    }

                    {
                        let imports = tree_shaking_module.imports();
                        let exports = tree_shaking_module.exports();

                        // 分析使用情况
                        for import in imports {
                            debug!("    - import: {:?}", &import);
                            self.analyze_import_statement(
                                tree_shaking_module_map,
                                tree_shaking_module_id,
                                import,
                            );
                        }

                        for export in exports {
                            debug!("    - export: {:?}", &export);
                            self.analyze_export_statement(
                                tree_shaking_module_map,
                                tree_shaking_module_id,
                                export,
                            );
                        }
                    };
                }
            } else {
                let module_graph = self.context.module_graph.read().unwrap();
                for (dep_id, _) in module_graph.get_dependencies(tree_shaking_module_id) {
                    let dep_tree_shake_module = tree_shaking_module_map.get_mut(dep_id);

                    if let Some(dep_tree_shake_module) = dep_tree_shake_module {
                        dep_tree_shake_module.used_exports = UsedExports::All;
                    }
                }
                drop(module_graph);
            }
            // 处理 dynamic import 的情况，把他们都设置成为具备副作用
            self.markup_module_by_resolve_type(tree_shaking_module_id, tree_shaking_module_map);
        }

        self.cleanup_no_used_export_module(&modules_to_remove);

        modules_to_remove
    }

    // Require & DynamicImport 的模块先标记为 UsedExports::All，后续可以优化
    fn markup_module_by_resolve_type(
        &self,
        tree_shaking_module_id: &ModuleId,
        tree_shaking_module_map: &mut HashMap<ModuleId, TreeShakingModule>,
    ) {
        let module_graph = self.context.module_graph.read().unwrap();
        for (dep, edge) in module_graph.get_dependencies(tree_shaking_module_id) {
            if matches!(edge.resolve_type, crate::module::ResolveType::DynamicImport) {
                let tree_shake_module = tree_shaking_module_map.get_mut(dep).unwrap();
                tree_shake_module.side_effects = true;
                tree_shake_module.used_exports = UsedExports::All;
            } else if matches!(edge.resolve_type, crate::module::ResolveType::Require) {
                let tree_shake_module = tree_shaking_module_map.get_mut(dep).unwrap();
                tree_shake_module.used_exports = UsedExports::All;
            }
        }
        drop(module_graph);
    }

    fn make_toposort(&self) -> (Vec<ModuleId>, Vec<ModuleId>, Vec<Vec<ModuleId>>) {
        let module_graph = self.context.module_graph.read().unwrap();
        let entry_modules = module_graph.get_entry_modules();
        let (sorted_modules, cycle_modules) = module_graph.toposort();
        (entry_modules, sorted_modules, cycle_modules)
    }

    fn markup_entry_modules_as_side_effects(&self, entry_modules: Vec<ModuleId>) {
        let mut module_graph = self.context.module_graph.write().unwrap();
        for entry_module_id in entry_modules {
            let module = module_graph.get_module_mut(&entry_module_id).unwrap();
            module.side_effects = true;
        }
    }
    #[allow(dead_code)]
    fn markup_cycle_modules_as_side_effects(&self, cycle_modules: Vec<Vec<ModuleId>>) {
        let mut module_graph = self.context.module_graph.write().unwrap();
        for cycle_module in cycle_modules {
            for module_id in cycle_module {
                let module = module_graph.get_module_mut(&module_id).unwrap();
                module.side_effects = true;
            }
        }
    }

    #[allow(dead_code)]
    fn cleanup_no_used_export_module(&self, modules_to_remove: &Vec<ModuleId>) {
        let mut module_graph = self.context.module_graph.write().unwrap();
        debug!("modules_to_remove: {:?}", &modules_to_remove);

        for module_id in modules_to_remove {
            module_graph.remove_module_and_deps(module_id);
        }
    }

    fn create_tree_shaking_module_map(
        &self,
        sorted_modules: Vec<crate::module::ModuleId>,
    ) -> (Vec<ModuleId>, HashMap<ModuleId, TreeShakingModule>) {
        let mut tree_shaking_module_ids = vec![];
        let mut tree_shaking_module_map = HashMap::new();
        for module_id in sorted_modules {
            let mut dependencies_of_module: Vec<ModuleId> = vec![];
            {
                let module_graph_r = self.context.module_graph.read().unwrap();
                dependencies_of_module.extend(
                    module_graph_r
                        .get_dependencies(&module_id)
                        .iter()
                        .map(|f| f.0.clone())
                        .collect::<Vec<_>>(),
                );
            }

            let mut module_graph = self.context.module_graph.write().unwrap();
            let module = module_graph.get_module_mut(&module_id).unwrap();

            // external 模块，设置为副作用模块
            if module.is_external() {
                module.side_effects = true;
            }

            if !module.get_module_type().is_script() {
                // 非 js 模块，以及他们的依赖，都设置为副作用模块
                for dep_module_id in dependencies_of_module {
                    let dep_module = module_graph.get_module_mut(&dep_module_id).unwrap();
                    dep_module.side_effects = true;
                }
            }

            let module = module_graph.get_module(&module_id).unwrap();
            let tree_shaking_module = TreeShakingModule::new(module);
            tree_shaking_module_ids.push(tree_shaking_module.id.clone());
            tree_shaking_module_map.insert(tree_shaking_module.id.clone(), tree_shaking_module);
        }
        (tree_shaking_module_ids, tree_shaking_module_map)
    }
}

#[cfg(test)]
mod tests {
    use mako_core::tokio;

    use crate::assert_display_snapshot;
    use crate::test_helper::{read_dist_file, setup_compiler};

    #[tokio::test(flavor = "multi_thread")]
    async fn test_tree_shaking_esm_cjs_mix() {
        let compiler = setup_compiler("test/build/tree-shaking-esm-mix-cjs", false);
        compiler.compile().unwrap();
        let content = read_dist_file(&compiler, "dist/index.js");

        assert!(content.contains(r#"cjs.js"#), "cjs.js missing in dist");
    }

    #[tokio::test(flavor = "multi_thread")]
    async fn test_tree_shaking_export_all() {
        let compiler = setup_compiler("test/build/tree-shaking-export-all", false);
        compiler.compile().unwrap();
        let content = read_dist_file(&compiler, "dist/index.js");

        assert!(!content.contains(r#"2.ts"#), "2.ts should be shaked");
    }

    #[tokio::test(flavor = "multi_thread")]
    async fn test_tree_shaking_esm_cjs_mix_2() {
        let compiler = setup_compiler("test/build/tree-shaking-esm-mix-cjs-2", false);
        compiler.compile().unwrap();
        let content = read_dist_file(&compiler, "dist/index.js");

        println!("{}", content);
        assert!(!content.contains(r#"cjs.js"#), "cjs.js should be shaked");
    }

    #[tokio::test(flavor = "multi_thread")]
    async fn test_tree_shaking() {
        let compiler = setup_compiler("test/build/tree-shaking", false);
        compiler.compile().unwrap();
        let content = read_dist_file(&compiler, "dist/index.js");
        assert_display_snapshot!(content);
    }

    #[tokio::test(flavor = "multi_thread")]
    async fn test_tree_shaking_reexport() {
        let compiler = setup_compiler("test/build/tree-shaking_reexport", false);
        compiler.compile().unwrap();
        let content = read_dist_file(&compiler, "dist/index.js");
        assert_display_snapshot!(content);
    }

    #[tokio::test(flavor = "multi_thread")]
    async fn test_tree_shaking_named_reexport() {
        let compiler = setup_compiler("test/build/tree-shaking_named_reexport", false);
        compiler.compile().unwrap();
        let content = read_dist_file(&compiler, "dist/index.js");
        assert_display_snapshot!(content);
    }

    #[tokio::test(flavor = "multi_thread")]
    async fn test_tree_shaking_export_namespace() {
        let compiler = setup_compiler("test/build/tree-shaking_export_namespace", false);
        compiler.compile().unwrap();
        let content = read_dist_file(&compiler, "dist/index.js");
        assert_display_snapshot!(content);
    }
    #[tokio::test(flavor = "multi_thread")]
    async fn test_tree_shaking_named_export() {
        let compiler = setup_compiler("test/build/tree-shaking_named_export", false);
        compiler.compile().unwrap();
        let content = read_dist_file(&compiler, "dist/index.js");
        assert_display_snapshot!(content);
    }
    #[tokio::test(flavor = "multi_thread")]
    async fn test_tree_shaking_fn() {
        let compiler = setup_compiler("test/build/tree-shaking_fn", false);
        compiler.compile().unwrap();
        let content = read_dist_file(&compiler, "dist/index.js");
        assert_display_snapshot!(content);
    }
    #[tokio::test(flavor = "multi_thread")]
    async fn test_tree_shaking_side_effect() {
        let compiler = setup_compiler("test/build/tree-shaking_side_effect", false);
        compiler.compile().unwrap();
        let content = read_dist_file(&compiler, "dist/index.js");
        assert_display_snapshot!(content);
    }
    #[tokio::test(flavor = "multi_thread")]
    async fn test_tree_shaking_class() {
        let compiler = setup_compiler("test/build/tree-shaking_class", false);
        compiler.compile().unwrap();
        let content = read_dist_file(&compiler, "dist/index.js");
        assert_display_snapshot!(content);
    }
    #[tokio::test(flavor = "multi_thread")]
    async fn test_tree_shaking_exported() {
        let compiler = setup_compiler("test/build/tree-shaking_exported", false);
        compiler.compile().unwrap();
        let content = read_dist_file(&compiler, "dist/index.js");
        assert_display_snapshot!(content);
    }
    #[tokio::test(flavor = "multi_thread")]
    async fn test_tree_shaking_export_default() {
        let compiler = setup_compiler("test/build/tree-shaking_export_default", false);
        compiler.compile().unwrap();
        let content = read_dist_file(&compiler, "dist/index.js");
        assert_display_snapshot!(content);
    }
    #[tokio::test(flavor = "multi_thread")]
    async fn test_tree_shaking_issues_271() {
        let compiler = setup_compiler("test/build/tree-shaking_issues_271", false);
        compiler.compile().unwrap();
        let content = read_dist_file(&compiler, "dist/index.js");
        assert_display_snapshot!(content);
    }
    #[tokio::test(flavor = "multi_thread")]
    async fn test_tree_shaking_jsx() {
        let compiler = setup_compiler("test/build/tree-shaking_jsx", false);
        compiler.compile().unwrap();
        let content = read_dist_file(&compiler, "dist/index.js");
        assert_display_snapshot!(content);
    }
    #[tokio::test(flavor = "multi_thread")]
    async fn test_tree_shaking_dynamic_import() {
        let compiler = setup_compiler("test/build/tree-shaking_dynamic-import", false);
        compiler.compile().unwrap();
        let content = read_dist_file(&compiler, "dist/index.js");
        assert_display_snapshot!(content);
        let content = read_dist_file(&compiler, "dist/a_ts-async.js");
        assert_display_snapshot!(content);
    }
    #[tokio::test(flavor = "multi_thread")]
    async fn test_tree_shaking_style() {
        let compiler = setup_compiler("test/build/tree-shaking_style", false);
        compiler.compile().unwrap();
        let content = read_dist_file(&compiler, "dist/index.js");
        assert_display_snapshot!(content);
    }
    #[tokio::test(flavor = "multi_thread")]
    async fn test_tree_shaking_require_esm() {
        let compiler = setup_compiler("test/build/tree-shaking_require_esm", false);
        compiler.compile().unwrap();
        let content = read_dist_file(&compiler, "dist/index.js");
        assert_display_snapshot!(content);
    }
    #[tokio::test(flavor = "multi_thread")]
    async fn test_tree_shaking_import_multi() {
        let compiler = setup_compiler("test/build/tree-shaking_import_multi", false);
        compiler.compile().unwrap();
        let content = read_dist_file(&compiler, "dist/index.js");
        assert_display_snapshot!(content);
    }
    #[tokio::test(flavor = "multi_thread")]
    async fn test_tree_shaking_import_self() {
        let compiler = setup_compiler("test/build/tree-shaking_import_self", false);
        compiler.compile().unwrap();
        let content = read_dist_file(&compiler, "dist/index.js");
        assert_display_snapshot!(content);
    }
}
