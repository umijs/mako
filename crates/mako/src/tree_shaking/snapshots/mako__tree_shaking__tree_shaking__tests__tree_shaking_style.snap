---
source: crates/mako/src/tree_shaking/tree_shaking.rs
expression: content
---
var chunksIdToUrlMap = {};
var cssChunksIdToUrlMap = {
    "index.tsx": "index.css"
};
var e = "index.tsx";
var cssInstalledChunks = {
    "index.tsx": 0
};
var m = {
    "foo.tsx": function(module, exports, require) {
        Object.defineProperty(exports, "__esModule", {
            value: true
        });
        function _export(target, all) {
            for(var name in all)Object.defineProperty(target, name, {
                enumerable: true,
                get: all[name]
            });
        }
        _export(exports, {
            a: function() {
                return a;
            },
            b: function() {
                return b;
            }
        });
        function a() {}
        function b() {}
    },
    "index.tsx": function(module, exports, require) {
        Object.defineProperty(exports, "__esModule", {
            value: true
        });
        "";
        require("foo.tsx");
        console.log('ff');
    }
};
function createRuntime(makoModules, entryModuleId) {
    var modulesRegistry = {};
    function requireModule(moduleId) {
        if (moduleId === '$$IGNORED$$') return {};
        var cachedModule = modulesRegistry[moduleId];
        if (cachedModule !== undefined) {
            if (cachedModule.error) {
                throw cachedModule.error;
            }
            return cachedModule.exports;
        }
        var module = {
            id: moduleId,
            exports: {}
        };
        modulesRegistry[moduleId] = module;
        try {
            var execOptions = {
                id: moduleId,
                module: module,
                factory: makoModules[moduleId],
                require: requireModule
            };
            requireModule.requireInterceptors.forEach(function(interceptor) {
                interceptor(execOptions);
            });
            execOptions.factory.call(execOptions.module.exports, execOptions.module, execOptions.module.exports, execOptions.require);
        } catch (e) {
            modulesRegistry[moduleId].error = e;
            throw e;
        }
        return module.exports;
    }
    requireModule.requireInterceptors = [];
    !(function() {
        requireModule.chunkEnsures = {};
        requireModule.ensure = function(chunkId) {
            return Promise.all(Object.keys(requireModule.chunkEnsures).reduce(function(promises, key) {
                requireModule.chunkEnsures[key](chunkId, promises);
                return promises;
            }, []));
        };
    })();
    !(function() {
        requireModule.jsonpInstalled = {};
        var installedChunks = requireModule.jsonpInstalled;
        requireModule.chunkEnsures.jsonp = function(chunkId, promises) {
            var data = installedChunks[chunkId];
            if (data === 0) return;
            if (data) {
                promises.push(data[2]);
            } else {
                var promise = new Promise(function(resolve, reject) {
                    data = installedChunks[chunkId] = [
                        resolve,
                        reject
                    ];
                });
                promises.push((data[2] = promise));
                var url = requireModule.publicPath + chunksIdToUrlMap[chunkId];
                var error = new Error();
                var onLoadEnd = function(event) {
                    data = installedChunks[chunkId];
                    if (data !== 0) installedChunks[chunkId] = undefined;
                    if (data) {
                        var errorType = event && event.type;
                        var src = event && event.target && event.target.src;
                        error.message = 'Loading chunk ' + chunkId + ' failed. (' + errorType + ' : ' + src + ')';
                        error.name = 'ChunkLoadError';
                        error.type = errorType;
                        data[1](error);
                    }
                };
                requireModule.loadScript(url, onLoadEnd, 'chunk-' + chunkId);
                return promise;
            }
        };
    })();
    !(function() {
        var inProgress = {};
        requireModule.loadScript = function(url, done, key) {
            if (inProgress[url]) {
                return inProgress[url].push(done);
            }
            var script = document.createElement('script');
            script.timeout = 120;
            script.src = url;
            inProgress[url] = [
                done
            ];
            var onLoadEnd = function(prev, event) {
                clearTimeout(timeout);
                var doneFns = inProgress[url];
                delete inProgress[url];
                if (script.parentNode) script.parentNode.removeChild(script);
                if (doneFns) {
                    doneFns.forEach(function(fn) {
                        return fn(event);
                    });
                }
                if (prev) return prev(event);
            };
            var timeout = setTimeout(onLoadEnd.bind(null, undefined, {
                type: 'timeout',
                target: script
            }), 120000);
            script.onerror = onLoadEnd.bind(null, script.onerror);
            script.onload = onLoadEnd.bind(null, script.onload);
            document.head.appendChild(script);
        };
    })();
    !(function() {
        requireModule.cssInstalled = cssInstalledChunks;
        requireModule.findStylesheet = function(url) {
            return Array.from(document.querySelectorAll('link[href][rel=stylesheet]')).find(function(link) {
                var linkUrl = link.getAttribute('href').split('?')[0];
                return linkUrl === url || linkUrl === requireModule.publicPath + url;
            });
        };
        requireModule.createStylesheet = function(chunkId, url, oldTag, resolve, reject) {
            var link = document.createElement('link');
            link.rel = 'stylesheet';
            link.type = 'text/css';
            link.href = url;
            link.onerror = link.onload = function(event) {
                link.onerror = link.onload = null;
                if (event.type === 'load') {
                    cssInstalledChunks[chunkId] = 0;
                    resolve();
                } else {
                    delete cssInstalledChunks[chunkId];
                    var errorType = event && event.type;
                    var realHref = event && event.target && event.target.href;
                    var err = new Error('Loading CSS chunk ' + chunkId + ' failed.\n(' + realHref + ')');
                    err.code = 'CSS_CHUNK_LOAD_FAILED';
                    err.type = errorType;
                    err.request = realHref;
                    link.parentNode.removeChild(link);
                    reject(err);
                }
            };
            if (oldTag) {
                oldTag.parentNode.insertBefore(link, oldTag.nextSibling);
            } else {
                document.head.appendChild(link);
            }
            return link;
        };
        requireModule.chunkEnsures.css = function(chunkId, promises) {
            if (cssInstalledChunks[chunkId]) {
                promises.push(cssInstalledChunks[chunkId]);
            } else if (cssInstalledChunks[chunkId] !== 0 && cssChunksIdToUrlMap[chunkId]) {
                cssInstalledChunks[chunkId] = new Promise(function(resolve, reject) {
                    var url = cssChunksIdToUrlMap[chunkId];
                    var fullUrl = requireModule.publicPath + url;
                    if (requireModule.findStylesheet(url)) {
                        resolve();
                    } else {
                        requireModule.createStylesheet(chunkId, fullUrl, null, resolve, reject);
                    }
                });
                promises.push(cssInstalledChunks[chunkId]);
                return promises;
            }
        };
    })();
    var jsonpCallback = function(data) {
        var installedChunks = requireModule.jsonpInstalled;
        var chunkIds = data[0];
        var modules = data[1];
        if (chunkIds.some(function(id) {
            return installedChunks[id] !== 0;
        })) {
            registerModules(modules);
        }
        for(var i = 0; i < chunkIds.length; i++){
            var id = chunkIds[i];
            if (installedChunks[id]) {
                installedChunks[id][0]();
            }
            installedChunks[id] = 0;
        }
    };
    var registerModules = function(modules) {
        for(var id in modules){
            makoModules[id] = modules[id];
        }
    };
    !function() {
        requireModule.publicPath = "/";
    }();
    !function() {
        registerModules({
            "@swc/helpers/_/_interop_require_default": function(module, exports, require) {
                Object.defineProperty(exports, "__esModule", {
                    value: true
                });
                function _export(target, all) {
                    for(var name in all)Object.defineProperty(target, name, {
                        enumerable: true,
                        get: all[name]
                    });
                }
                _export(exports, {
                    _interop_require_default: function() {
                        return _interop_require_default;
                    },
                    _: function() {
                        return _interop_require_default;
                    }
                });
                function _interop_require_default(obj) {
                    return obj && obj.__esModule ? obj : {
                        default: obj
                    };
                }
            },
            "@swc/helpers/_/_interop_require_wildcard": function(module, exports, require) {
                Object.defineProperty(exports, "__esModule", {
                    value: true
                });
                function _export(target, all) {
                    for(var name in all)Object.defineProperty(target, name, {
                        enumerable: true,
                        get: all[name]
                    });
                }
                _export(exports, {
                    _interop_require_wildcard: function() {
                        return _interop_require_wildcard;
                    },
                    _: function() {
                        return _interop_require_wildcard;
                    }
                });
                function _getRequireWildcardCache(nodeInterop) {
                    if (typeof WeakMap !== "function") return null;
                    var cacheBabelInterop = new WeakMap();
                    var cacheNodeInterop = new WeakMap();
                    return (_getRequireWildcardCache = function(nodeInterop) {
                        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
                    })(nodeInterop);
                }
                function _interop_require_wildcard(obj, nodeInterop) {
                    if (!nodeInterop && obj && obj.__esModule) return obj;
                    if (obj === null || typeof obj !== "object" && typeof obj !== "function") return {
                        default: obj
                    };
                    var cache = _getRequireWildcardCache(nodeInterop);
                    if (cache && cache.has(obj)) return cache.get(obj);
                    var newObj = {};
                    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
                    for(var key in obj)if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
                        var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
                        if (desc && (desc.get || desc.set)) Object.defineProperty(newObj, key, desc);
                        else newObj[key] = obj[key];
                    }
                    newObj.default = obj;
                    if (cache) cache.set(obj, newObj);
                    return newObj;
                }
            },
            "@swc/helpers/_/_export_star": function(module, exports, require) {
                Object.defineProperty(exports, "__esModule", {
                    value: true
                });
                function _export(target, all) {
                    for(var name in all)Object.defineProperty(target, name, {
                        enumerable: true,
                        get: all[name]
                    });
                }
                _export(exports, {
                    _export_star: function() {
                        return _export_star;
                    },
                    _: function() {
                        return _export_star;
                    }
                });
                function _export_star(from, to) {
                    Object.keys(from).forEach(function(k) {
                        if (k !== "default" && !Object.prototype.hasOwnProperty.call(to, k)) Object.defineProperty(to, k, {
                            enumerable: true,
                            get: function() {
                                return from[k];
                            }
                        });
                    });
                    return from;
                }
            }
        });
    }();
    var exports = requireModule(entryModuleId);
    return {
        exports: exports,
        requireModule: requireModule,
        _modulesRegistry: modulesRegistry,
        _jsonpCallback: jsonpCallback,
        _makoModuleHotUpdate: requireModule.applyHotUpdate
    };
}
var runtime = createRuntime(m, e);
var root = typeof globalThis !== 'undefined' ? globalThis : self;
root.jsonpCallback = runtime._jsonpCallback;
root.modulesRegistry = runtime._modulesRegistry;
root.makoModuleHotUpdate = runtime._makoModuleHotUpdate;

//# sourceMappingURL=index.js.map
